sycl/source/detail/platform_impl.cpp

```diff
@@ -140,14 +141,28 @@ std::vector<platform> platform_impl::get_platforms() {
 // This function matches devices in the order of backend, device_type, and
 // device_num.
 static void filterDeviceFilter(std::vector<RT::PiDevice> &PiDevices,
-                               const plugin &Plugin) {
+                               RT::PiPlatform Platform) {
   device_filter_list *FilterList = SYCLConfig<SYCL_DEVICE_FILTER>::get();
   if (!FilterList)
     return;
+  std::vector<plugin> &Plugins = GlobalHandler::instance().getPlugins();
+  if (Plugins.size() == 0)
+    RT::initialize();
```

Comment:
Under what circumstances this check could fail?

Reply:
if platform::get_devices() is called first time, plugins are initialized and that's the only time RT will initialize plugins.
If there is a second call to platform::get_devices(), getPlugins() will return non-empty list.

Reply:
Ok, then this function/check doesn't look thread safe:
1.  `platform_impl::get_devices`(caller) doesn't have any locks
2.  `RT::initialize` modifies content of the `Plugins`,
3. Another thread may access Plugins in any intermediate state.

Also, if this function is being executed by multiple threads calls to `getStartingDeviceId` and `setLastDeviceId` can cause raises.

Reply:
I will add a lock for global plugins.

Reply:
Done.